File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.*;

class Solution {
    static PriorityQueue<Integer> pQ;
    public int[] solution(String[] operations) {
        int[] answer = new int[2];
        pQ = new PriorityQueue<>();

        StringTokenizer st;
        for (int i = 0; i < operations.length; i++) {
            st = new StringTokenizer(operations[i]);
            String oper = st.nextToken();
            int num = Integer.parseInt(st.nextToken());

            if(oper.equals("I")){
                pQ.add(num);
            }
            else {
                removeNum(num);
            }
        }

        if(pQ.isEmpty()) {
            answer[0] = 0;
            answer[1] = 0;
        }
        else{
            PriorityQueue<Integer> pQ_rev = new PriorityQueue<>(Comparator.reverseOrder());
            for(Integer num : pQ) pQ_rev.add(num);
            answer[0] = pQ_rev.remove();
            answer[1] = pQ.remove();
        }
       return answer;
    }

    static void removeNum(int removeFlag){
        if(!pQ.isEmpty()){
            //최댓값 삭제
            if(removeFlag==1){
                PriorityQueue<Integer> pQ_rev = new PriorityQueue<>(Comparator.reverseOrder());
                for(Integer num : pQ) pQ_rev.add(num);
                int maxNum = pQ_rev.remove();
                pQ.remove(maxNum);
            }
            //최소값 삭제
            else{
                pQ.remove();
            }
        }
    }
}

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.py
import heapq
min_heap = []

def deleteNum(flag):
    if len(min_heap) > 0:
        if flag == -1: # 최솟값 삭제
            heapq.heappop(min_heap)
        else: 
            min_heap.sort()
            min_heap.pop() # 최댓값 삭제
            
def solution(operations):
    answer = []
    for operation in operations:
        oper, num = operation.split()
        num = int(num)
        if oper == 'I':
            heapq.heappush(min_heap, num)
        else:
            deleteNum(num)
            
    min_heap.sort()
    if min_heap:
        answer.append(min_heap[-1])
        answer.append(min_heap[0])
    else:
        answer.append(0)
        answer.append(0)

    return answer